File: app/src/main/java/com/alphadevelopmentsolutions/frcscout/Fragments/TeamListFragment.java
package com.alphadevelopmentsolutions.frcscout.Fragments;

import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.LinearLayout;

import com.alphadevelopmentsolutions.frcscout.Adapters.FragmentViewPagerAdapter;
import com.alphadevelopmentsolutions.frcscout.Adapters.TeamListRecyclerViewAdapter;
import com.alphadevelopmentsolutions.frcscout.Classes.Match;
import com.alphadevelopmentsolutions.frcscout.Classes.Team;
import com.alphadevelopmentsolutions.frcscout.Enums.AllianceColor;
import com.alphadevelopmentsolutions.frcscout.R;

import java.util.ArrayList;

/**
 * A simple {@link Fragment} subclass.
 * Activities that contain this fragment must implement the
 * {@link TeamListFragment.OnFragmentInteractionListener} interface
 * to handle interaction events.
 * Use the {@link TeamListFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class TeamListFragment extends MasterFragment
{
    static final String ARG_ALLIANCE_COLOR = "ALLIANCE_COLOR";

    private String allianceColorString;

    private OnFragmentInteractionListener mListener;

    public TeamListFragment()
    {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @return A new instance of fragment TeamListFragment.
     */
    // TODO: Rename and change types and number of parameters
    public static TeamListFragment newInstance(@Nullable Match match, @Nullable AllianceColor allianceColor)
    {
        TeamListFragment fragment = new TeamListFragment();
        Bundle args = new Bundle();
        args.putString(ARG_MATCH_JSON, toJson(match));
        args.putString(ARG_ALLIANCE_COLOR, (allianceColor == null) ? AllianceColor.NONE.name() : allianceColor.name());
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);

        //check if any args were passed, specifically for team and match json
        if (getArguments() != null)
        {
            matchJson = getArguments().getString(ARG_MATCH_JSON);
            allianceColorString = getArguments().getString(ARG_ALLIANCE_COLOR);
        }

        loadTeamsThread = new Thread(new Runnable()
        {
            @Override
            public void run()
            {
                joinLoadingThread();

                if(match != null && allianceColorString != null && !allianceColorString.equals(""))
                    allianceColor = AllianceColor.getColorFromString(allianceColorString);

                //get all teams at event
                teams = event.getTeams(null, null, database);

                //if a match and alliance color was specified,
                //remove any teams that are not in that match or alliance color
                if(match != null && allianceColor != null)
                {
                    for(Team team : new ArrayList<>(teams))
                    {
                        //team not in match / alliance color
                        if(match.getTeamAllianceColor(team) != allianceColor)
                            teams.remove(team);
                    }

                    //add all the teams to the searchedTeams arraylist
                    searchedTeams = new ArrayList<>(teams);
                }
                else
                    //add all the teams to the searchedTeams arraylist
                    searchedTeams = new ArrayList<>(teams);
            }
        });

        loadTeamsThread.start();
    }

    private RecyclerView teamsRecyclerView;

    private EditText teamSearchEditText;

    private ArrayList<Team> teams;
    private ArrayList<Team> searchedTeams;

    private Thread loadTeamsThread;

    private AllianceColor allianceColor;

    private LinearLayout allianceViewPagerLinearLayout;
    private TabLayout allianceTabLayout;
    private ViewPager allianceViewPager;

    private Toolbar searchTeamsToolbar;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState)
    {
        // Inflate the layout for this fragment
        View view =  inflater.inflate(R.layout.fragment_team_list, container, false);

        //gets rid of the shadow on the actionbar
        context.dropActionBar();

        teamsRecyclerView = view.findViewById(R.id.TeamsRecyclerView);
        teamSearchEditText = view.findViewById(R.id.TeamSearchEditText);
        allianceViewPagerLinearLayout = view.findViewById(R.id.AllianceViewPagerLinearLayout);
        allianceTabLayout = view.findViewById(R.id.AllianceTabLayout);
        allianceViewPager = view.findViewById(R.id.AllianceViewPager);
        searchTeamsToolbar = view.findViewById(R.id.SearchTeamsToolbar);



        //join back up with the load teams thread
        try
        {
            loadTeamsThread.join();
        }
        catch (InterruptedException e)
        {
            e.printStackTrace();
        }

        context.setTitle((match == null) ? event.getName() : match.toString());

        //hide search bar if match is specified
        if(match != null)
            searchTeamsToolbar.setVisibility(View.GONE);

        if(match == null || (allianceColor == AllianceColor.BLUE || allianceColor == AllianceColor.RED))
        {
            final TeamListRecyclerViewAdapter teamListRecyclerViewAdapter = new TeamListRecyclerViewAdapter(match, searchedTeams, context);

            teamsRecyclerView.setLayoutManager(new LinearLayoutManager(context));
            teamsRecyclerView.setAdapter(teamListRecyclerViewAdapter);

            teamSearchEditText.addTextChangedListener(new TextWatcher()
            {
                @Override
                public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2)
                {

                }

                @Override
                public void onTextChanged(CharSequence searchText, int start, int before, int count)
                {
                    //You only need to reset the list if you are removing from your search, adding the objects back
                    if (count < before)
                    {
                        //Reset the list
                        for (int i = 0; i < teams.size(); i++)
                        {
                            Team team = teams.get(i);

                            //check if the contact doesn't exist in the viewable list
                            if (!searchedTeams.contains(team))
                            {
                                //add it and notify the recyclerview
                                searchedTeams.add(i, team);
                                teamListRecyclerViewAdapter.notifyItemInserted(i);
                                teamListRecyclerViewAdapter.notifyItemRangeChanged(i, searchedTeams.size());
                            }

                        }
                    }

                    //Delete from the list
                    for (int i = 0; i < searchedTeams.size(); i++)
                    {
                        Team team = searchedTeams.get(i);
                        String name = team.getId() + " - " + team.getName();

                        //If the contacts name doesn't equal the searched name
                        if (!name.toLowerCase().contains(searchText.toString().toLowerCase()))
                        {
                            //remove it from the list and notify the recyclerview
                            searchedTeams.remove(i);
                            teamListRecyclerViewAdapter.notifyItemRemoved(i);
                            teamListRecyclerViewAdapter.notifyItemRangeChanged(i, searchedTeams.size());

                            //this prevents the index from passing the size of the list,
                            //stays on the same index until you NEED to move to the next one
                            i--;
                        }
                    }
                }

                @Override
                public void afterTextChanged(Editable editable)
                {

                }
            });
        }

        //if match specified, setup the viewpager and hide the recyclerview
        else
        {
            allianceViewPagerLinearLayout.setVisibility(View.VISIBLE);
            teamsRecyclerView.setVisibility(View.GONE);

            FragmentViewPagerAdapter viewPagerAdapter = new FragmentViewPagerAdapter(getChildFragmentManager());

            viewPagerAdapter.addFragment(TeamListFragment.newInstance(match, AllianceColor.BLUE), getString(R.string.blue_alliance));
            viewPagerAdapter.addFragment(TeamListFragment.newInstance(match, AllianceColor.RED), getString(R.string.red_alliance));

            allianceViewPager.setAdapter(viewPagerAdapter);
            allianceTabLayout.setupWithViewPager(allianceViewPager);
        }

        return view;
    }

    // TODO: Rename method, update argument and hook method into UI event
    public void onButtonPressed(Uri uri)
    {
        if (mListener != null)
        {
            mListener.onFragmentInteraction(uri);
        }
    }

    @Override
    public void onAttach(Context context)
    {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener)
        {
            mListener = (OnFragmentInteractionListener) context;
        } else
        {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach()
    {
        super.onDetach();
        mListener = null;
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnFragmentInteractionListener
    {
        // TODO: Update argument type and name
        void onFragmentInteraction(Uri uri);
    }
}

File: app/src/main/java/com/alphadevelopmentsolutions/frcscout/Fragments/TeamListFragment.kt
package com.alphadevelopmentsolutions.frcscout.Fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.SearchView
import androidx.recyclerview.widget.LinearLayoutManager
import com.alphadevelopmentsolutions.frcscout.Adapters.FragmentViewPagerAdapter
import com.alphadevelopmentsolutions.frcscout.Adapters.TeamListRecyclerViewAdapter
import com.alphadevelopmentsolutions.frcscout.Classes.Tables.Match
import com.alphadevelopmentsolutions.frcscout.Classes.Tables.Team
import com.alphadevelopmentsolutions.frcscout.Enums.AllianceColor
import com.alphadevelopmentsolutions.frcscout.R
import kotlinx.android.synthetic.main.fragment_team_list.view.*
import java.util.*

class TeamListFragment : MasterFragment()
{
    override fun onBackPressed(): Boolean
    {
        return false
    }

    private var allianceColorString: String? = null

    private var teams: ArrayList<Team>? = null
    private var searchedTeams: ArrayList<Team>? = null

    private var loadTeamsThread: Thread? = null

    private var allianceColor: AllianceColor? = null

    private var previousSearchLength: Int = 0

    private lateinit var teamListAdapter: TeamListRecyclerViewAdapter

    override fun onCreate(savedInstanceState: Bundle?)
    {
        super.onCreate(savedInstanceState)

        //check if any args were passed, specifically for team and match json
        if (arguments != null)
        {
            matchJson = arguments!!.getString(ARG_MATCH_JSON)
            allianceColorString = arguments!!.getString(ARG_ALLIANCE_COLOR)
        }

        loadTeamsThread = Thread(Runnable {
            loadingThread.join()

            if (match != null && allianceColorString != null && allianceColorString != "")
                allianceColor = AllianceColor.getColorFromString(allianceColorString!!)

            //get all teams at event
            teams = event!!.getTeams(null, null, database)

            //if a match and alliance color was specified,
            //remove any teams that are not in that match or alliance color
            searchedTeams = if (match != null && allianceColor != null)
            {
                for (team in ArrayList(teams!!))
                {
                    //team not in match / alliance color
                    if (match!!.getTeamAllianceColor(team) != allianceColor)
                        teams!!.remove(team)
                }

                //add all the teams to the searchedTeams arraylist
                ArrayList(teams!!)
            } else
                //add all the teams to the searchedTeams arraylist
                ArrayList(teams!!)

            teamListAdapter = TeamListRecyclerViewAdapter(match, searchedTeams!!, context)
        })

        loadTeamsThread!!.start()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View?
    {
        // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_team_list, container, false)

        //gets rid of the shadow on the actionbar
        context.dropActionBar()

        loadTeamsThread!!.join()

        context.setToolbarTitle(if (match == null) context.getString(R.string.teams) else match.toString())
        context.isSearchViewVisible = match == null
        context.isToolbarScrollable = match == null

        view.AllianceTabLayout.setBackgroundColor(context.primaryColor)

        if (match == null || allianceColor == AllianceColor.BLUE || allianceColor == AllianceColor.RED)
        {
            Thread(Runnable{

                val layoutManager = LinearLayoutManager(context)

                context.runOnUiThread{

                    view.TeamsRecyclerView.layoutManager = layoutManager
                    view.TeamsRecyclerView.adapter = teamListAdapter

                    isLoading = false
                }

            }).start()

            if(match == null)
            {
                context.setSearchViewOnTextChangeListener(object: SearchView.OnQueryTextListener{
                    override fun onQueryTextSubmit(p0: String?): Boolean
                    {
                        return false
                    }

                    override fun onQueryTextChange(searchText: String?): Boolean
                    {
                        val searchLength = searchText?.length ?: 0

                        //You only need to reset the list if you are removing from your search, adding the objects back
                        if (searchLength < previousSearchLength)
                        {
                            //Reset the list
                            for (i in teams!!.indices)
                            {
                                val team = teams!![i]

                                //check if the contact doesn't exist in the viewable list
                                if (!searchedTeams!!.contains(team))
                                {
                                    //add it and notify the recyclerview
                                    searchedTeams!!.add(i, team)
                                    teamListAdapter.notifyItemInserted(i)
                                    teamListAdapter.notifyItemRangeChanged(i, searchedTeams!!.size)
                                }

                            }
                        }

                        //Delete from the list
                        var i = 0
                        while (i < searchedTeams!!.size)
                        {
                            val team = searchedTeams!![i]
                            val name = team.id.toString() + " - " + team.name

                            //If the contacts name doesn't equal the searched name
                            if (!name.toLowerCase().contains(searchText.toString().toLowerCase()))
                            {
                                //remove it from the list and notify the recyclerview
                                searchedTeams!!.removeAt(i)
                                teamListAdapter.notifyItemRemoved(i)
                                teamListAdapter.notifyItemRangeChanged(i, searchedTeams!!.size)

                                //this prevents the index from passing the size of the list,
                                //stays on the same index until you NEED to move to the next one
                                i--
                            }
                            i++
                        }

                        previousSearchLength = searchLength

                        return false
                    }

                })
            }

        }

        else
        {
            context.setToolbarTitle(match.toString())
            view.AllianceViewPagerLinearLayout.visibility = View.VISIBLE
            view.TeamsRecyclerView.visibility = View.GONE

            val viewPagerAdapter = FragmentViewPagerAdapter(childFragmentManager)

            val blueAllianceFrag = newInstance(match, AllianceColor.BLUE)
            val redAllianceFrag = newInstance(match, AllianceColor.RED)

            viewPagerAdapter.addFragment(blueAllianceFrag, context.getString(R.string.blue_alliance))
            viewPagerAdapter.addFragment(redAllianceFrag, context.getString(R.string.red_alliance))

            view.AllianceViewPager.adapter = viewPagerAdapter
            view.AllianceTabLayout.setupWithViewPager(view.AllianceViewPager)
            view.AllianceTabLayout.setSelectedTabIndicatorColor(context.primaryColorDark)

            context.lockDrawerLayout(true, View.OnClickListener { context.onBackPressed() })

        }//if match specified, setup the viewpager and hide the recyclerview

        return super.onCreateView(view, !(match == null || allianceColor == AllianceColor.BLUE || allianceColor == AllianceColor.RED))
    }

    override fun onPause()
    {
        super.onPause()
        if(context.isSearchViewVisible)
            context.isSearchViewVisible = false
    }

    override fun onResume()
    {
        super.onResume()
        if(!context.isSearchViewVisible && match == null)
            context.isSearchViewVisible = true

        teamListAdapter.notifyItemChanged(0)
    }

    override fun onDestroyView()
    {
        if (match != null)
            context.unlockDrawerLayout()

        if(context.isSearchViewVisible)
            context.isSearchViewVisible = false

        super.onDestroyView()
    }

    override fun onDestroy()
    {
        super.onDestroy()
        context.setSearchViewOnTextChangeListener(null)
    }

    companion object
    {
        internal const val ARG_ALLIANCE_COLOR = "ALLIANCE_COLOR"

        /**
         * Creates a new instance
         * @param match to get teams from
         * @param allianceColor to get teams from
         * @return A new instance of fragment [TeamListFragment].
         */
        fun newInstance(match: Match?, allianceColor: AllianceColor?): TeamListFragment
        {
            val fragment = TeamListFragment()
            val args = Bundle()
            args.putString(ARG_MATCH_JSON, toJson(match))
            args.putString(ARG_ALLIANCE_COLOR, allianceColor?.name ?: AllianceColor.NONE.name)
            fragment.arguments = args
            return fragment
        }
    }
}
